parquetschema/autoschema/gen.go

```diff
@@ -0,0 +1,399 @@
+package autoschema
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"strings"
+	"time"
+
+	"github.com/fraugster/parquet-go/parquet"
+	"github.com/fraugster/parquet-go/parquetschema"
+)
+
+// GenerateSchema auto-generates a schema definition for a provided object's type
+// using reflection. The generated schema is meant to be compatible with
+// github.com/fraugster/parquet-go/floor's reflection-based marshalling/unmarshalling.
+func GenerateSchema(obj interface{}) (*parquetschema.SchemaDefinition, error) {
+	valueObj := reflect.ValueOf(obj)
+	columns, err := generateSchema(valueObj.Type())
+	if err != nil {
+		return nil, fmt.Errorf("can't generate schema: %w", err)
+	}
+
+	return &parquetschema.SchemaDefinition{
+		RootColumn: &parquetschema.ColumnDefinition{
+			SchemaElement: &parquet.SchemaElement{
+				Name: "autogen_schema",
+			},
+			Children: columns,
+		},
+	}, nil
+}
+
+func generateSchema(objType reflect.Type) ([]*parquetschema.ColumnDefinition, error) {
+	if objType.Kind() == reflect.Ptr {
+		objType = objType.Elem()
+	}
+
+	if objType.Kind() != reflect.Struct {
+		return nil, errors.New("can't generate schema: provided object needs to be of type struct or *struct")
+	}
+
+	columns := []*parquetschema.ColumnDefinition{}
+
+	for i := 0; i < objType.NumField(); i++ {
+		fieldType := objType.Field(i)
+		fieldName := fieldNameToLower(fieldType)
+
+		column, err := generateField(fieldType.Type, fieldName)
+		if err != nil {
+			return nil, err
+		}
+
+		columns = append(columns, column)
+	}
+
+	return columns, nil
+}
+
+func generateField(fieldType reflect.Type, fieldName string) (*parquetschema.ColumnDefinition, error) {
+	switch fieldType.Kind() {
+	case reflect.Bool:
+		return &parquetschema.ColumnDefinition{
+			SchemaElement: &parquet.SchemaElement{
+				Type:           parquet.TypePtr(parquet.Type_BOOLEAN),
+				Name:           fieldName,
+				RepetitionType: parquet.FieldRepetitionTypePtr(parquet.FieldRepetitionType_REQUIRED),
+			},
+		}, nil
+	case reflect.Int:
+		return &parquetschema.ColumnDefinition{
+			SchemaElement: &parquet.SchemaElement{
+				Type:           parquet.TypePtr(parquet.Type_INT32),
```

Comment:
are we sure int should map to int32 and not int64?

Reply:
Changed to int64.